src/Raven.Client/Documents/BulkInsert/BulkInsertOperation.cs

```diff
@@ -423,12 +495,19 @@ private async Task HandleErrors(string documentId, Exception e)
             await ThrowOnUnavailableStream(documentId, e).ConfigureAwait(false);
         }
 
-        private IDisposable ConcurrencyCheck()
+        private IDisposable _streamLockDisposable;
+        private async ValueTask<IDisposable> ConcurrencyCheckAsync()
         {
             if (Interlocked.CompareExchange(ref _concurrentCheck, 1, 0) == 1)
                 throw new InvalidOperationException("Bulk Insert store methods cannot be executed concurrently.");
 
-            return new DisposableAction(() => Interlocked.CompareExchange(ref _concurrentCheck, 0, 1));
+            _streamLockDisposable = await _streamLock.WriterLockAsync().ConfigureAwait(false);
+
+            return new DisposableAction(() =>
```

Comment:
return here dedicated readonly struct that on dispose does the same thing

Reply:
done